File: 404.sum-of-left-leaves.py
#
# @lc app=leetcode id=404 lang=python3
#
# [404] Sum of Left Leaves
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root) -> int:
        return self.sum_left(root, False)

    def sum_left(self, root, is_left) -> int:
        if not root:
            return 0
        if not root.left and not root.right and is_left:  # Check if it's a left leaf node
            return root.val
        return self.sum_left(root.left, True) + self.sum_left(root.right, False)
        
        
# @lc code=end

File: 404.sum-of-left-leaves.java
/*
 * @lc app=leetcode id=404 lang=java
 *
 * [404] Sum of Left Leaves
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumOfLeftLeaves(TreeNode root) {
        return sumLeft(root, false);
    }

    public int sumLeft(TreeNode root, boolean isLeft) {
        if (root == null) {
            return 0;
        }

        if (isLeft && root.left == null && root.right == null) {
            return root.val;
        }
        return sumLeft(root.left, true) + sumLeft(root.right, false);
    }
}
// @lc code=end

